CodingConventions.md

```diff
@@ -0,0 +1,310 @@
+later, you need to revisit the _entire_ business logic again.
+
+
+# Appendix
+
+## Engines (interface deprecated)
+We model the data flow inside a node as a 'data flow graph'. An `Engine` is the deprecated interface for a vertex within the node's data flow graph. 
+New vertices should implement the [`Component` interface](https://github.com/onflow/flow-go/blob/57f89d4e96259f08fe84163c91ecd32484401b45/module/component/component.go#L22)
+and throw unexpected errors using the related [irrecoverable context](https://github.com/onflow/flow-go/blob/277b6515add6136946913747efebd508f0419a25/module/irrecoverable/irrecoverable.go).
+
+See [related tech debt issue](https://github.com/dapperlabs/flow-go/issues/6361) for more information on the deprecation.
+
+Generally, we consider node-internal components as _trusted_ and external nodes as untrusted data sources. 
+The `Engine` API differentiates between trusted and untrusted inputs:
+
+* **Trusted inputs** from other components within the same node are handed to the `Engine` through
+  ```golang
+  // SubmitLocal submits an event originating on the local node.
+  SubmitLocal(event interface{})
+
+  // ProcessLocal processes an event originating on the local node.
+  ProcessLocal(event interface{}) error
+  ```
+* **Untrusted inputs** from other nodes are handed to the `Engine` via
+  ```golang
+  // Submit submits the given event from the node with the given origin ID
+  // for processing in a non-blocking manner. It returns instantly and logs
+  // a potential processing error internally when done.
+  Submit(channel Channel, originID flow.Identifier, event interface{})
+
+  // Process processes the given event from the node with the given origin ID
+  // in a blocking manner. It returns the potential processing error when
+  // done.
+  Process(channel Channel, originID flow.Identifier, event interface{}) error
+  ```
+
+Furthermore, engines offer synchronous and asynchronous processing of their inputs:
+
+* Methods `engine.ProcessLocal` and `engine.Process` run their logic in the caller's routine. Therefore, the core
+  business logic can return errors, which we should capture and propagate.
+* Methods `engine.SubmitLocal` and `engine.Submit` run the business logic asynchronously. (no error return)
+    * Note: computationally very cheap and predominantly non-blocking computation should _still_ be executed in the
+      calling routine. This is beneficial to avoid a cascade of go-routine launches when stacking multiple engines.
+
+### Guidelines
+
+* In the error handling within an `Engine`, differentiate between trusted and untrusted inputs. For example, most
+  engines expect specific input types. Inputs with incompatible type should result in a specific sentinel error (
+  e.g. `InvalidMessageType`).    
+  For messages that come from the network (methods `Submit` and `Process`), we expect invalid types. But if we get an
+  invalid message from a trusted local engine
```

Comment:
```suggestion
  invalid message type from a trusted local engine
```

Maybe we should be more specific about what is considered to be invalid. 

For instance, the follower engine could receive a block from the network or synchronization engine. 

If a block from sync engine to follower engine is considered to be trusted input, then it's still possible for the block to be invalid. Sync engine doesn't validate the the message (block), it's just responsible for delivering the message.

Also whether or not a message is "invalid" might depend on another module's state, for instance, a new block at the moment is not invalid, but after a while might become "invalid" because it's conflicting with the finalized block. 

Maybe we could restrict the message type here, saying if we get an invalid message _type_ from a trusted local engine, then something is broken.

Reply:
👍 Good call-out. I will update this to be more detailed. 

Personally though I don't think we should constrain the requirement to "invalid message _type_". Different components will have different definitions for what is a valid message. In all cases, an unexpected message type is invalid. But in many cases, an expected type with malformed data is also invalid. 

We also should be moving away from non-type-safe interfaces between components (engines), so the invalid type case will be less of a concern over time. 

In the case of the sync engine and follower engine, the sync engine is responsible for forwarding blocks, but not verifying them. So a `SyncedBlock` message containing an invalid `flow.Block` would be a valid internal message from the `sync` engine. This is a bit of a special case IMO where an external message is forwarded without validation.

But to take another example, suppose HotStuff notifies Engine X about some block B being finalized. HotStuff guarantees that finalized blocks can be queried from the database. If Engine X cannot read B from the database, then the notification from HotStuff could be considered an invalid message and a critical failure.

Reply:
Ah I see this section is within the `Engine`-specific appendix, so I'll take your suggestion directly for this section, and add some more detail above: https://github.com/onflow/flow-go/pull/2575/commits/1d1106f74c9f3e18f2a6eb18fa69ae1986f13c2f